docs/current/guides/snippets/build-test-publish-php/dagger.php

```diff
@@ -0,0 +1,277 @@
+<?php
+// include auto-loader
+include 'vendor/autoload.php';
+
+use GraphQL\Client;
+
+class DaggerPipeline {
+
+  private $client;
+
+  // constructor
+  public function __construct() {
+    // initialize client with
+    // endpoint from environment
+    $sessionPort = getenv('DAGGER_SESSION_PORT') or throw new Exception("DAGGER_SESSION_PORT doesn't exist");
+    $sessionToken = getenv('DAGGER_SESSION_TOKEN') or throw new Exception("DAGGER_SESSION_TOKEN doesn't exist");
+    $this->client = new Client(
+      'http://127.0.0.1:' . $sessionPort . '/query',
+      ['Authorization' => 'Basic ' . base64_encode($sessionToken . ':')]
+    );
+  }
+
+  // build base image
+  public function buildBaseImage() {
+    // get host working directory
+    $sourceQuery = <<<QUERY
+    query {
+      host {
+        directory (path: ".", exclude: ["vendor", "ci"]) {
+          id
+        }
+      }
+    }
+    QUERY;
+    $sourceDir = $this->executeQuery($sourceQuery);
+
+    // build runtime image
+    // install tools and PHP extensions
+    // configure Apache webserver root and rewriting
+    $runtimeQuery = <<<QUERY
+    query {
+      container {
+        from(address: "php:8.2-apache-buster") {
```

Comment:
Can we extract this PHP version into a constant at the top of the file? This often gets bumped, and putting it deep in the code makes that difficult. I personally add a comment with a link to where I look when I want to update this, e.g. `// https://hub.docker.com/_/php/tags?page=1&name=apache-buster`

Reply:
Done, although I used a class variable as interpolating constants into a heredoc query is not straightforward.

Reply:
That sounds the better solution in the PHP context.